File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n 
  | n <= 0    = Nothing
  | n > 64    = Nothing
  | otherwise = Just (go 1 1)
  where
    go x k
      | k >= n    = x
      | otherwise = go (2 * x) (k + 1)   

total :: Integer
total = sum (catMaybesList [square x | x <- [1..64]])
  where
    catMaybesList [] = []
    catMaybesList (Nothing : xs) = catMaybesList xs
    catMaybesList (Just x  : xs) = x : catMaybesList xs
  
      
    

File: solutions/c/grains/1/grains.c
#include "grains.h"


uint64_t square(uint8_t index) {
  uint64_t results[64] = {1};
  for (int i = 1; i < index; i++) {
    results[i] = 2 * results[i - 1];
  }
  return results[index - 1];
}


uint64_t total(void) {
  return square(64);
}
